pygmt/clib/session.py

```diff
@@ -1609,6 +1609,79 @@ def virtualfile_from_data(  # noqa: PLR0912
 
         return file_context
 
+    @contextlib.contextmanager
+    def virtualfile_out(
+        self, kind: Literal["dataset", "grid"] = "dataset", fname: str | None = None
+    ):
+        """
+        Create a virtual file or an actual file for storing output data.
+
+        If ``fname`` is not given, a virtual file will be created to store the output
+        data into a GMT data container and the function yields the name of the virutal
+        file. Otherwise, the output data will be written into the specified file and the
+        function simply yields the actual file name.
+
+        Parameters
+        ----------
+        kind
+            The data kind of the virtual file to create. Valid values are ``"dataset"``
+            and ``"grid"``. Ignored if ``fname`` is specified.
+        fname
+            The name of the actual file to write the output data. No virtual file will
+            be created.
+
+        Yields
+        ------
+        vfile : str
+            Name of the virtual file or the actual file.
+
+        Examples
+        --------
+        >>> from pathlib import Path
+        >>> from pygmt.clib import Session
+        >>> from pygmt.datatypes import _GMT_DATASET, _GMT_GRID
+        >>> from pygmt.helpers import GMTTempFile
+        >>>
+        >>> # Create a virtual file for storing the output table.
+        >>> with GMTTempFile(suffix=".txt") as tmpfile:
+        ...     with open(tmpfile.name, mode="w") as fp:
+        ...         print("1.0 2.0 3.0 TEXT", file=fp)
+        ...     with Session() as lib:
+        ...         with lib.virtualfile_out(kind="dataset") as vouttbl:
+        ...             lib.call_module("read", f"{tmpfile.name} {vouttbl} -Td")
+        ...             ds = lib.read_virtualfile(vouttbl, kind="dataset")
+        >>> isinstance(ds.contents, _GMT_DATASET)
+        True
+        >>>
+        >>> # Create a virtual file for storing the output grid.
+        >>> with Session() as lib:
+        ...     with lib.virtualfile_out(kind="grid") as voutgrd:
+        ...         lib.call_module("read", f"@earth_relief_01d_g {voutgrd} -Tg")
+        ...         outgrd = lib.read_virtualfile(voutgrd, kind="grid")
+        >>> isinstance(outgrd.contents, _GMT_GRID)
+        True
+        >>>
+        >>> # Write data to file without creating a virtual file
+        >>> with GMTTempFile(suffix=".nc") as tmpfile:
+        ...     with Session() as lib:
+        ...         with lib.virtualfile_out(
+        ...             kind="grid", fname=tmpfile.name
+        ...         ) as voutgrd:
+        ...             lib.call_module("read", f"@earth_relief_01d_g {voutgrd} -Tg")
+        ...             assert voutgrd == tmpfile.name
+        ...             assert Path(voutgrd).stat().st_size > 0
```

Comment:
Should we move this into a new `test_clib_virtualfile_out.py` file? Two reasons:

1. There might be other data types like [`GMT_IMAGE`](https://docs.generic-mapping-tools.org/6.5/devdocs/api.html#struct-image) and [`GMT_CUBE`](https://docs.generic-mapping-tools.org/6.5/devdocs/api.html#gmt-cube) in the future, and the doctest would get long.
2. Might want to test the output of the virtualfiles more properly using `pygmt.info` or `pygmt.grdinfo`.

Reply:
> Should we move this into a new `test_clib_virtualfile_out.py` file? Two reasons:
> 
> 1. There might be other data types like [`GMT_IMAGE`](https://docs.generic-mapping-tools.org/6.5/devdocs/api.html#struct-image) and [`GMT_CUBE`](https://docs.generic-mapping-tools.org/6.5/devdocs/api.html#gmt-cube) in the future, and the doctest would get long.

The main purpose of these doctests are to show how to use this method and the *outpu* virtual file in other modules. So I think we can only keep the `dataset` one only and remove all others.

> 2. Might want to test the output of the virtualfiles more properly using `pygmt.info` or `pygmt.grdinfo`.

Actually we can't. Because the virtual file is an output (i.e., `GMT_OUT`), it can't be passed as an *input* to `info`/`grdinfo`. I.e., `lib.call_module("info", f"{vouttbl}")` won't work.

Reply:
> So I think we can only keep the `dataset` one only and remove all others.

You can move the grid one to `test_clib_virtualfile_out.py`

> > 3\. Might want to test the output of the virtualfiles more properly using `pygmt.info` or `pygmt.grdinfo`.
> 
> Actually we can't. Because the virtual file is an output (i.e., `GMT_OUT`), it can't be passed as an _input_ to `info`/`grdinfo`. I.e., `lib.call_module("info", f"{vouttbl}")` won't work.

Ah ok, maybe not for the virtualfile then. But we should test that saving to a (non-virtual) file works, and try loading from there (i.e. roundtrip). Again, can put this in `test_clib_virtualfile_out.py`

Reply:
> > Should we move this into a new `test_clib_virtualfile_out.py` file? Two reasons:
> > 
> > 1. There might be other data types like [`GMT_IMAGE`](https://docs.generic-mapping-tools.org/6.5/devdocs/api.html#struct-image) and [`GMT_CUBE`](https://docs.generic-mapping-tools.org/6.5/devdocs/api.html#gmt-cube) in the future, and the doctest would get long.
> 
> The main purpose of these doctests are to show how to use this method and the _outpu_ virtual file in other modules. So I think we can only keep the `dataset` one only and remove all others.

In commit 9108917d5cd73bbc8897937636e8a3252ff7e50f, I've removed the doctest for `grid` and kept the doctests for `dataset` and `fname`. I also realized that these doctests are almost the same as the ones in `read_virtualfile`. This makes sense, because `read_virtualfile` was implemented to read the data from an *output* virtualfile. So, I also simplify the doctests in `read_virtualfile` using the newly added `virtualfile_out` method.

In other words, we probably need to move the doctests of `read_virtualfile` into a separate test file in the future.

Reply:
> > So I think we can only keep the `dataset` one only and remove all others.
> 
> You can move the grid one to `test_clib_virtualfile_out.py`

As mentioned in the above comment, the grid one is already doctested in the `read_virtualfile` method.

> > > 3. Might want to test the output of the virtualfiles more properly using `pygmt.info` or `pygmt.grdinfo`.
> > 
> > 
> > Actually we can't. Because the virtual file is an output (i.e., `GMT_OUT`), it can't be passed as an _input_ to `info`/`grdinfo`. I.e., `lib.call_module("info", f"{vouttbl}")` won't work.
> 
> Ah ok, maybe not for the virtualfile then. But we should test that saving to a (non-virtual) file works (when `fname` is used), and try loading from there (i.e. roundtrip). Again, can put this in `test_clib_virtualfile_out.py`

Done in 888f1c9c96b4de496aa2b80368c70fab45ccc1ea.